core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java

```diff
@@ -122,12 +309,32 @@ default PPLTypeChecker getTypeChecker() {
   private final Map<BuiltinFunctionName, List<Pair<CalciteFuncSignature, FunctionImp>>>
       externalFunctionRegistry;
 
-  private PPLFuncImpTable(Builder builder) {
+  /**
+   * The registry for built-in agg functions. Agg Functions defined by the PPL specification, whose
+   * implementations are independent of any specific data storage, should be registered here
+   * internally.
+   */
+  private final ImmutableMap<BuiltinFunctionName, AggHandler> aggFunctionRegistry;
+
+  /**
+   * The external agg function registry. Agg Functions whose implementations depend on a specific
+   * data engine should be registered here. This reduces coupling between the core module and
+   * particular storage backends.
+   */
+  private final Map<BuiltinFunctionName, AggHandler> aggExternalFunctionRegistry;
+
+  private PPLFuncImpTable(Builder builder, AggBuilder aggBuilder) {
     final ImmutableMap.Builder<BuiltinFunctionName, List<Pair<CalciteFuncSignature, FunctionImp>>>
         mapBuilder = ImmutableMap.builder();
     builder.map.forEach((k, v) -> mapBuilder.put(k, List.copyOf(v)));
     this.functionRegistry = ImmutableMap.copyOf(mapBuilder.build());
     this.externalFunctionRegistry = new HashMap<>();
+
+    final ImmutableMap.Builder<BuiltinFunctionName, AggHandler> aggMapBuilder =
+        ImmutableMap.builder();
+    aggBuilder.map.forEach(aggMapBuilder::put);
+    this.aggFunctionRegistry = ImmutableMap.copyOf(aggMapBuilder.build());
+    this.aggExternalFunctionRegistry = new HashMap<>();
```

Comment:
Both `externalFunctionRegistry` and `aggExternalFunctionRegistry` should change to `CurrentHashMap` or add `synchronized` keyword in `registerExternalFunction` and `registerExternalAggFunction`. Because the `PPLFuncImpTable` instance is a static member. No problem for now since we only register external functions when initialize `OpenSearchExecutionEngine`. But it's a public API, we'd better protect it.

Reply:
Already make two map to `ConCurrentHashMap` and slightly change the code of `registerExternalFunction`. The put in `registerExternalAggFunction` is already safe. Please check it.